components/script/dom/domparser.rs

```diff
@@ -39,19 +43,27 @@ impl DOMParser {
 }
 
 impl<'a> DOMParserMethods for JSRef<'a, DOMParser> {
+    // http://domparsing.spec.whatwg.org/#the-domparser-interface
     fn ParseFromString(self,
-                       _s: DOMString,
+                       s: DOMString,
                        ty: DOMParserBinding::SupportedType)
                        -> Fallible<Temporary<Document>> {
-        let window = self.window.root();
-        //FIXME: these should probably be FromParser when we actually parse the string (#3756).
+        let window = self.window.root().clone();
+        let url = Some(window.get_url());
+        let content_type = DOMParserBinding::SupportedTypeValues::strings[ty as uint].to_string();
         match ty {
             Text_html => {
-                Ok(Document::new(*window, None, HTMLDocument, Some("text/html".to_string()),
-                                 NotFromParser))
+                let document = Document::new(window, url.clone(), HTMLDocument,
+                                             Some(content_type), FromParser).root().clone();
+                let parser = ServoHTMLParser::new(url.clone(), document).root().clone();
+                parser.parse_chunk(s);
+                parser.finish();
+                document.set_ready_state(DocumentReadyStateValues::Complete);
+                Ok(Temporary::from_rooted(document))
             }
             Text_xml => {
-                Ok(Document::new(*window, None, NonHTMLDocument, Some("text/xml".to_string()),
+                //FIXME: this should probably be FromParser when we actually parse the string (#3756).
```

Comment:
@ChrisParis  Why fixme here?

Reply:
A `Document` keeps track of whether it was created from a parsed stream, or from some other source. This comment was meant to say that when the stub `Document` created in the next line is replaced by a call to an actual XML parser, then `NotFromParser` should be replaced by `FromParser`, to property indicate the source. I see that in master, there is now a call to an XML parser, so I would think the `Document` should be marked as `FromParser`.

Reply:
@ChrisParis  Thanks! Could you explain more about this?
> I see that in master, there is now a call to an XML parser, so I would think the Document should be marked as FromParser.

So, How do you suggest to solve "fixme"?

Reply:
In the code statement after the comment, replace `DocumentSource::NotFromParser` with `DocumentSource::FromParser`, and delete the fixme comment. I've been away from this for three years, so this will need another set of eyes on it.

Reply:
@ChrisParis  Thanks.  :D